earthly/python/Earthfile

```diff
@@ -23,22 +27,30 @@ python-base:
     # Stop python code running in the container saving pointless .pyc files.
     ENV PYTHONDONTWRITEBYTECODE=1
 
-    # Install Poetry to manage python.
+    # Install Poetry using the recommended installer.
     RUN curl -sSL https://install.python-poetry.org | python3 -
-    # Prevent warning when running on machines with more than 6 cores.
-    #   "Connection pool is full, discarding connection: pypi.org. Connection pool size: 10"
+
+    # Adjust Poetry's configuration to prevent connection pool warnings.
     RUN poetry config installer.max-workers 10
 
-# Common python builder setup.
-# Parameters:
-#   opts      : Options passed to the `poetry install` command.
+    # Install ruff for linting instead of flake8.
+    RUN pip3 install ruff
+
+# Utilize the POETRY_SETUP UDC for setting up Poetry within our environment.
+poetry:
+    FROM +python-base
+    # The setup steps for POETRY_SETUP would be included here, assuming it's defined elsewhere.
+    # This placeholder assumes the necessary Poetry and project setup is performed here.
+
 BUILDER:
     FUNCTION
-    ARG opts
-
-    # Copy our dependencies.
+    FROM +poetry
```

Comment:
Linting should not be part of the build, and this change to BUILDER will break all our python builds that use it.

There needs to be a `CHECK:` FUNCTION which uses the `builder`.
It needs to:
1. Validate the Ruff configuration in the project matches our fixed configuration (not done here yet).
2. Run ruff on the src

Thats all, it doesn't need to run poetry at all, and it shouldn't even assume we use poetry because we do have python code which needs linting but which is not a poetry package.  Such as our python based build scripts.

Then in those targets we can add a `check:` target which should just look like this:

```
check:
    COPY --dir ./src .
    DO ./../../earthly/python+CHECK
```

The CHECK function should look something like this (untested):
```
CHECK:
    FUNCTION

    FROM +python-base

    COPY .ruff.toml .

    RUN /scripts/std_checks.py
```

See: `examples/rust/Earthfile` for an example of using a `std_check.py` file to do the python checks (in this case its just validate the config file and run `ruff check .` but we could add more checks to this file in future.

You will need to make a `std_checks.py` and include it in the builder in `/scripts` so that it can be run by the CHECK function.

Reply:
Add on to what Steven said the `CHECK` function should be available for other repo to use too for example, this spelling check https://github.com/input-output-hk/catalyst-ci/blob/master/earthly/cspell/Earthfile is used in `cat-voice` https://github.com/input-output-hk/catalyst-voices/blob/main/Earthfile

Reply:
thank you! will look at them asap